File: packages/contracts-python/src/aegis_contracts/primitives.py
"""Canonical identifiers, timestamps, and ordering primitives."""

from __future__ import annotations

import re
from collections.abc import Callable
from datetime import UTC, datetime
from typing import Annotated, Any

from pydantic import BeforeValidator, Field, PlainSerializer

from aegis_contracts.errors import ContractErrorCode, ContractValidationError

AUTHORED_ID_PATTERN = re.compile(
    r"^(asset|incident|alert|evidence|agent-session|business-unit|edge|scenario|"
    r"scenario-version|relationship|service|user|device|identity|database|control):"
    r"[a-z0-9][a-z0-9._-]{0,126}$"
)
RUNTIME_ID_PATTERN = re.compile(
    r"^(evt|run|trc|inc|alt|evd|ags|prp|apr|act|mdl|scr|hyp)_[0-9A-HJKMNP-TV-Z]{26}$"
)

Sequence = Annotated[int, Field(ge=0)]
Revision = Annotated[int, Field(ge=0)]


def _validate_authored_id(value: Any) -> str:
    if not isinstance(value, str):
        raise ContractValidationError(
            code=ContractErrorCode.INVALID_IDENTIFIER,
            message="Authored identifier must be a string",
            details={"value": value},
        )
    if not AUTHORED_ID_PATTERN.fullmatch(value):
        raise ContractValidationError(
            code=ContractErrorCode.INVALID_IDENTIFIER,
            message=f"Invalid authored identifier: {value}",
            details={"value": value},
        )
    return value


def _validate_runtime_id(prefix: str, value: Any) -> str:
    if not isinstance(value, str):
        raise ContractValidationError(
            code=ContractErrorCode.INVALID_IDENTIFIER,
            message="Runtime identifier must be a string",
            details={"value": value},
        )
    if not RUNTIME_ID_PATTERN.fullmatch(value):
        raise ContractValidationError(
            code=ContractErrorCode.INVALID_IDENTIFIER,
            message=f"Invalid runtime identifier: {value}",
            details={"value": value, "expectedPrefix": prefix},
        )
    if not value.startswith(f"{prefix}_"):
        raise ContractValidationError(
            code=ContractErrorCode.INVALID_IDENTIFIER,
            message=f"Runtime identifier must use prefix {prefix}_",
            details={"value": value, "expectedPrefix": prefix},
        )
    return value


def _make_runtime_id_validator(prefix: str) -> Callable[[Any], str]:
    def validator(value: Any) -> str:
        return _validate_runtime_id(prefix, value)

    return validator


def _validate_utc_timestamp(value: Any) -> datetime:
    if isinstance(value, datetime):
        parsed = value
    elif isinstance(value, str):
        normalized = value.replace("Z", "+00:00")
        try:
            parsed = datetime.fromisoformat(normalized)
        except ValueError as exc:
            raise ContractValidationError(
                code=ContractErrorCode.VALIDATION_FAILED,
                message=f"Invalid UTC timestamp: {value}",
                details={"value": value},
            ) from exc
    else:
        raise ContractValidationError(
            code=ContractErrorCode.VALIDATION_FAILED,
            message="Timestamp must be an ISO-8601 string or datetime",
            details={"value": value},
        )

    if parsed.tzinfo is None:
        raise ContractValidationError(
            code=ContractErrorCode.VALIDATION_FAILED,
            message="Timestamp must be timezone-aware (UTC)",
            details={"value": value},
        )

    return parsed.astimezone(UTC)


def _serialize_utc_timestamp(value: datetime) -> str:
    return value.astimezone(UTC).isoformat().replace("+00:00", "Z")


def validate_authored_id(value: Any) -> str:
    return _validate_authored_id(value)


def validate_runtime_id(prefix: str, value: Any) -> str:
    return _validate_runtime_id(prefix, value)


AuthoredId = Annotated[str, BeforeValidator(_validate_authored_id)]
AssetId = AuthoredId
IncidentId = Annotated[str, BeforeValidator(_validate_authored_id)]
AlertId = Annotated[str, BeforeValidator(_validate_authored_id)]
EvidenceId = Annotated[str, BeforeValidator(_validate_authored_id)]
AgentSessionId = Annotated[str, BeforeValidator(_validate_authored_id)]
ScenarioId = Annotated[str, BeforeValidator(_validate_authored_id)]
RelationshipId = AuthoredId
ClusterId = AuthoredId
EdgeId = AuthoredId

EventId = Annotated[str, BeforeValidator(_make_runtime_id_validator("evt"))]
RunId = Annotated[str, BeforeValidator(_make_runtime_id_validator("run"))]
TraceId = Annotated[str, BeforeValidator(_make_runtime_id_validator("trc"))]
CausationId = Annotated[str, BeforeValidator(_make_runtime_id_validator("evt"))]
CorrelationId = Annotated[str, BeforeValidator(_make_runtime_id_validator("trc"))]
ProposalId = Annotated[str, BeforeValidator(_make_runtime_id_validator("prp"))]
ApprovalId = Annotated[str, BeforeValidator(_make_runtime_id_validator("apr"))]
ActionId = Annotated[str, BeforeValidator(_make_runtime_id_validator("act"))]
ModelId = Annotated[str, BeforeValidator(_make_runtime_id_validator("mdl"))]
HypothesisId = Annotated[str, BeforeValidator(_make_runtime_id_validator("hyp"))]

UtcTimestamp = Annotated[
    datetime,
    BeforeValidator(_validate_utc_timestamp),
    PlainSerializer(_serialize_utc_timestamp, return_type=str),
]
SimTimestamp = UtcTimestamp

File: packages/contracts-ts/src/primitives.ts
import { z } from 'zod';

import { ContractErrorCode, ContractValidationError } from './errors';
import { SUPPORTED_SCHEMA_VERSIONS } from './versioning';

const AUTHORED_ID_PATTERN =
  /^(asset|incident|alert|evidence|agent-session|business-unit|edge|scenario|scenario-version|relationship|service|user|device|identity|database|control):[a-z0-9][a-z0-9._-]{0,126}$/;

const RUNTIME_ID_PATTERN =
  /^(evt|run|trc|inc|alt|evd|ags|prp|apr|act|mdl|scr|hyp|gen)_[0-9A-HJKMNP-TV-Z]{26}$/;

const UTC_TIMESTAMP_PATTERN = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{1,6})?Z$/;

export function validateAuthoredId(value: string): string {
  if (!AUTHORED_ID_PATTERN.test(value)) {
    throw new ContractValidationError({
      code: ContractErrorCode.INVALID_IDENTIFIER,
      message: `Invalid authored identifier: ${value}`,
      details: { value },
    });
  }
  return value;
}

export function validateRuntimeId(prefix: string, value: string): string {
  if (!RUNTIME_ID_PATTERN.test(value) || !value.startsWith(`${prefix}_`)) {
    throw new ContractValidationError({
      code: ContractErrorCode.INVALID_IDENTIFIER,
      message: `Invalid runtime identifier: ${value}`,
      details: { value, expectedPrefix: prefix },
    });
  }
  return value;
}

export const authoredIdSchema = z.string().superRefine((value, ctx) => {
  try {
    validateAuthoredId(value);
  } catch (error) {
    if (error instanceof ContractValidationError) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: error.message });
    }
  }
});

export const assetIdSchema = authoredIdSchema;
export const incidentIdSchema = authoredIdSchema;
export const alertIdSchema = authoredIdSchema;
export const evidenceIdSchema = authoredIdSchema;
export const agentSessionIdSchema = authoredIdSchema;
export const scenarioIdSchema = authoredIdSchema;
export const clusterIdSchema = authoredIdSchema;
export const edgeIdSchema = authoredIdSchema;

export const eventIdSchema = z.string().superRefine((value, ctx) => {
  try {
    validateRuntimeId('evt', value);
  } catch (error) {
    if (error instanceof ContractValidationError) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: error.message });
    }
  }
});

export const runIdSchema = z.string().superRefine((value, ctx) => {
  try {
    validateRuntimeId('run', value);
  } catch (error) {
    if (error instanceof ContractValidationError) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: error.message });
    }
  }
});

export const traceIdSchema = z.string().superRefine((value, ctx) => {
  try {
    validateRuntimeId('trc', value);
  } catch (error) {
    if (error instanceof ContractValidationError) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: error.message });
    }
  }
});

export const causationIdSchema = eventIdSchema;
export const correlationIdSchema = traceIdSchema;
export const proposalIdSchema = z.string().superRefine((value, ctx) => {
  try {
    validateRuntimeId('prp', value);
  } catch (error) {
    if (error instanceof ContractValidationError) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: error.message });
    }
  }
});
export const approvalIdSchema = z.string().superRefine((value, ctx) => {
  try {
    validateRuntimeId('apr', value);
  } catch (error) {
    if (error instanceof ContractValidationError) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: error.message });
    }
  }
});
export const actionIdSchema = z.string().superRefine((value, ctx) => {
  try {
    validateRuntimeId('act', value);
  } catch (error) {
    if (error instanceof ContractValidationError) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: error.message });
    }
  }
});
export const modelIdSchema = z.string().superRefine((value, ctx) => {
  try {
    validateRuntimeId('mdl', value);
  } catch (error) {
    if (error instanceof ContractValidationError) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: error.message });
    }
  }
});
export const hypothesisIdSchema = z.string().superRefine((value, ctx) => {
  try {
    validateRuntimeId('hyp', value);
  } catch (error) {
    if (error instanceof ContractValidationError) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: error.message });
    }
  }
});
export const generationRequestIdSchema = z.string().superRefine((value, ctx) => {
  try {
    validateRuntimeId('gen', value);
  } catch (error) {
    if (error instanceof ContractValidationError) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, message: error.message });
    }
  }
});

export const utcTimestampSchema = z.string().superRefine((value, ctx) => {
  if (!UTC_TIMESTAMP_PATTERN.test(value)) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: `Invalid UTC timestamp: ${value}`,
    });
    return;
  }
  const parsed = Date.parse(value);
  if (Number.isNaN(parsed)) {
    ctx.addIssue({
      code: z.ZodIssueCode.custom,
      message: `Invalid UTC timestamp: ${value}`,
    });
  }
});

export const simTimestampSchema = utcTimestampSchema;
export const sequenceSchema = z.number().int().min(0);
export const revisionSchema = z.number().int().min(0);

export function assertSupportedSchemaVersion(
  contractName: string,
  schemaVersion: number,
  traceId?: string,
): void {
  const supported = SUPPORTED_SCHEMA_VERSIONS[contractName];
  if (!supported) {
    throw new ContractValidationError({
      code: ContractErrorCode.VALIDATION_FAILED,
      message: `Unknown contract name: ${contractName}`,
      details: { contractName },
      traceId,
    });
  }
  if (!supported.includes(schemaVersion)) {
    throw new ContractValidationError({
      code: ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
      message: `Unsupported schema version ${String(schemaVersion)} for contract ${contractName}`,
      details: {
        contractName,
        schemaVersion,
        supportedVersions: supported,
      },
      traceId,
    });
  }
}
